src/utils.ts

```diff
@@ -538,34 +538,62 @@ export async function getDelegatesBySpace(
   const spaceIn = ['', space];
   if (space.includes('.eth')) spaceIn.push(space.replace('.eth', ''));
 
+  type Delegation = {
+    delegator: string;
+    delegate: string;
+    space: string;
+    timestamp: number;
+  };
   const PAGE_SIZE = 1000;
-  let result = [];
-  let page = 0;
+  let result: Delegation[] = [];
   const params: any = {
     delegations: {
       __args: {
         where: {
           space_in: spaceIn
         },
         first: PAGE_SIZE,
-        skip: 0
+        skip: 0,
+        orderBy: 'timestamp',
+        orderDirection: 'asc'
       },
       delegator: true,
       space: true,
-      delegate: true
+      delegate: true,
+      timestamp: true
     }
   };
   if (snapshot !== 'latest') {
     params.delegations.__args.block = { number: snapshot };
   }
 
+  const isSameDelegate = (a: Delegation, b: Delegation) => {
+    return (
+      a.delegator === b.delegator &&
+      a.delegate === b.delegate &&
+      a.space === b.space
+    );
+  };
+
   while (true) {
-    params.delegations.__args.skip = page * PAGE_SIZE;
+    const pivot = result[result.length - 1];
+    params.delegations.__args.where.timestamp_gte = pivot?.timestamp || 0;
+
+    let pageDelegations: Delegation[] =
+      (await subgraphRequest(subgraphUrl, params)).delegations || [];
+
+    // Removing duplicates results between pages because of the _gte filter
+    if (pivot) {
+      // List of delegation that can potentially be duplicates in the new page results
+      const pivotDelegations = result.filter((d) => isSameDelegate(d, pivot));
+
+      pageDelegations = pageDelegations.filter(
+        (d) => !pivotDelegations.some((p) => isSameDelegate(p, d))
+      );
```

Comment:
```suggestion
      pageDelegations = pageDelegations.filter(
         (d) => !result.some((p) => isSameDelegate(p, d))
      );
```

Reply:
the `pivotDelegations` is the potential duplicates, so if we use timestamp_gte = 1 on page 1, when we got to page 2, `pivotDelegations` will only contain delegate with `timestamp` == 1.

It was added to decrease the number of search loops, so if `pivotDelegations` is 2 elements, we have 2000 loop (2 * 1000).

By removing it, we will end up with 1000 * 1000 on page 2, and 2000 * 1000 on page 3, etc ...

Reply:
The best solution will be to implement a search loop which stop searching after the first negative return, since the array is sorted by timestamp, which will drop our loop count to single digit most of the time.

This will make the code less readable, so I go the route of computation is cheaper than dev time for maintenance

Reply:
No I think i am not able to explain it correctly 😄 but i mean `pivotDelegations` can just be `1` record, but there can be multiple duplicates in new page because we using `timestamp_gte`

Reply:
Maybe i am confused but Imagine if the first page returns
```JSON
[{delegate: "A", timestamp: 1},{delegate: "B", timestamp: 2},{delegate: "C", timestamp: 2}]
```
second page returns 
```JSON
[{delegate: "B", timestamp: 2},{delegate: "C", timestamp: 2},{delegate: "D", timestamp: 3}]
```

so now `pivotDelegations` will be `[{delegate: "C", timestamp: 2}]`, but it should be `[{delegate: "B", timestamp: 2},{delegate: "C", timestamp: 2}]`

Reply:
Maybe we should filter with just `timestamp `

```javascript
const pivotDelegations = result.filter((d) => isSameTimestamp(d, pivot));

pageDelegations = pageDelegations.filter(
  (d) => !pivotDelegations.some((p) => isSameTimestamp(p, d))
);
```

Reply:
You're absolutely correct, the pivotDelegation filter should not use the `isSameDelegate`, but another filter to extract all delegates with same timestamp

Reply:
Code refactored, should be more easier to understand

Logic extracted to delegation.ts, and split into multiple functions for easier reading and testing